stream-chat-android-client/src/main/java/io/getstream/chat/android/client/events/caching/TypingEventCache.kt

```diff
@@ -0,0 +1,234 @@
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    val currentlyTypingUsers: Map<TypingEventKeyValue, SelfStoppingTypingEvent> = _currentlyTypingUsers
+
+    /**
+     * Processes the incoming chat event accordingly.
+     *
+     * @param event The chat event to be processed.
+     */
+    override fun processEvent(event: ChatEvent) {
+        when (event) {
+            is TypingStartEvent -> {
+                processTypingStartEvent(event)
+            }
+            is TypingStopEvent -> {
+                processTypingStopEvent(event)
+            }
+            else -> {
+                throw IllegalArgumentException(
+                    "This class should be only used to cache typing events. " +
+                        "All other events will remain unprocessed."
+                )
+            }
+        }
+    }
+
+    /**
+     * Processes chat events by stopping the old "cleaning" job held
+     * by the previous value stored in the map under the same key
+     * and starts a new timed "cleaning" job.
+     *
+     * @param typingStartEvent The typing start event to be processed.
+     */
+    private fun processTypingStartEvent(typingStartEvent: TypingStartEvent) {
+        val keyValue = typingStartEvent.toStartTypingEventKeyValue()
+
+        // Cancel the previous timed death
+        _currentlyTypingUsers.getOrDefault(keyValue, null)?.cancelJob()
+
+        _currentlyTypingUsers[keyValue] = SelfStoppingTypingEvent(
+            coroutineScope = coroutineScope,
+            typingStartEvent = typingStartEvent,
+            delayTimeMs = cleanStaleEventMs
+        ) { typingStopEvent -> processTypingStopEvent(typingStopEvent) }
+
+        // Signal that the event has been processed
+        onEventFired(typingStartEvent)
+    }
+
+    /**
+     * Cancels the old "cleaning" job held by the previous
+     * value stored in the map under the same key and removes it.
+     *
+     * @param typingStopEvent The typing stop event to be processed.
+     */
+    private fun processTypingStopEvent(typingStopEvent: TypingStopEvent) {
+        val keyValue = typingStopEvent.toStartTypingEventKeyValue()
+
+        // Cancel the previous timed death
+        _currentlyTypingUsers.getOrDefault(keyValue, null)?.cancelJob()
+
+        _currentlyTypingUsers.remove(keyValue)
+
+        // Signal that the event has been processed
+        onEventFired(typingStopEvent)
+    }
+
+    /**
+     * Cancels [coroutineScope].
+     */
+    fun cancel() {
+        coroutineScope.cancel()
+    }
+}
+
+/**
+ * Used to provide key value equality for distinct typing events.
+ * A single value such as a [User] is not a good qualifier
+ * because a single user could type in multiple channels across
+ * the time span of a few seconds.
+ *
+ * @param user The user currently the typing event is tied to.
+ * @param cid The channel the the typing event is tied to.
+ */
+@InternalStreamChatApi
+internal data class TypingEventKeyValue(
```

Comment:
this is a very strange name for a class. Maybe TypingEventData?

Also - do we even need this class? both of the typing events (start, stop) have the same data as this, and we could just destructure it or format a string instead to use as a key. Something like: `userId:cid` can be the key.

Because it doesn't make sense - in the code, you transform both the Stop and Start events using `toStartTypingEventKeyValue`, which is confusing.

I'd just rather have a function:

```
fun typingEventToKey(user, cid) = "${user.id}:$cid"
```
or something,  than create a new class that does confusing transformations and then is used (as an object) as a key. Strings are much more flexible when it comes to keys and we can store any kind of information in them, nicely formatted, while being easier on the memory (roughly).